MoreLinq/EndsWith.cs

```diff
@@ -55,16 +59,68 @@ public static bool EndsWith<T>(this IEnumerable<T> first, IEnumerable<T> second)
         ///     This is the <see cref="IEnumerable{T}" /> equivalent of <see cref="string.EndsWith(string)" /> and
         ///     it calls <see cref="IEqualityComparer{T}.Equals(T,T)" /> on pairs of elements at the same index.
         /// </remarks>
+        [SuppressMessage("ReSharper", "InvokeAsExtensionMethod", Justification = "Avoid ambiguity with Reverse(IEnumerable<T>).")]
+        [SuppressMessage("ReSharper", "PossibleMultipleEnumeration", Justification = "Count() will use the concrete implementation of Length/Count, if available.")]
         public static bool EndsWith<T>(this IEnumerable<T> first, IEnumerable<T> second, IEqualityComparer<T> comparer)
         {
             if (first == null) throw new ArgumentNullException("first");
             if (second == null) throw new ArgumentNullException("second");
             if (comparer == null) comparer = EqualityComparer<T>.Default;
 
-            using (var firstIter = first.Reverse().GetEnumerator())
+            IList<T> firstList, secondList;
+            if ((firstList = first as IList<T>) != null &&
+                (secondList = second as IList<T>) != null)
+            {
+                return ListImpl(comparer, firstList, secondList);
+            }
+
+            // LinkedList does not implement IList<T>
+            LinkedList<T> firstLinkedList, secondLinkedList;
+            if ((firstLinkedList = first as LinkedList<T>) != null &&
+                (secondLinkedList = second as LinkedList<T>) != null)
+            {
+                return LinkedListImpl(comparer, firstLinkedList, secondLinkedList);
+            }
+
+            // Fall back to enumerating and comparing last N items, where N is the length of second.
+            // In order to properly dispose enumerator for second and only enumerate once,
+            // we must make a full copy of second to a list.
+            var newSecondList = second.ToList();
+            using (var firstIter = first.TakeLast(newSecondList.Count).GetEnumerator())
+            {
+                return newSecondList.All(item => firstIter.MoveNext() && comparer.Equals(firstIter.Current, item));
+            }
+        }
+
+        private static bool LinkedListImpl<T>(IEqualityComparer<T> comparer, LinkedList<T> firstLinkedList,
+            LinkedList<T> secondLinkedList)
+        {
+            using (var firstIter = LinkedListEx.Reverse(firstLinkedList).GetEnumerator())
+            using (var secondIter = LinkedListEx.Reverse(secondLinkedList).GetEnumerator())
```

Comment:
Similar to `ListImpl` below, if we know this is a `LinkedList`, we might as well iterate backwards from the `Last` property.